torch/csrc/jit/codegen/cuda/kernel_cache.cpp

```diff
@@ -437,6 +463,251 @@ std::vector<at::Tensor> FusionExecutorCache::runFusionWithInputs(
+            ival.isTensor(), "Cannot output non-tensor objects from a fusion.");
+        return ival.toTensor();
+      });
+
+  return fusion_output_tensors;
+}
+
+const std::vector<FusionSegmentRuntime::SchedulerEntryPtr>&
+FusionSegmentRuntime::schedulers() {
+  return heuristics_->heuristics();
+}
+
+namespace {
+using HashType = FusionSegmentRuntime::HashType;
+// Use a slightly more nontrivial combine to avoid collision
+//  (from Boost)
+inline HashType combineHash(HashType a, HashType b) {
+  return a ^
+      (b + 0x9e3779b9 + (a << 6) + // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+       (a >> 2)); // NOLINT(cppcoreguidelines-avoid-magic-numbers)
+}
+} // namespace
+
+FusionSegmentRuntime::HashType FusionSegmentRuntime::getHash(
+    SegmentHeuristics* sh) {
+  HashType h = 0;
+  for (auto& se_pt : sh->heuristics()) {
+    h = combineHash(h, SchedulerEntryHash()(*se_pt));
+  }
+  return h;
+}
+
+FusionSegmentRuntime::EntryTag::EntryTag(SegmentHeuristics* sh) {
+  heuristics_ = sh;
+  hash_ = FusionSegmentRuntime::getHash(sh);
+}
+
+bool FusionSegmentRuntime::EntryTag::operator==(
+    const FusionSegmentRuntime::EntryTag& other) const {
+  if (heuristics_->heuristics().size() !=
+      other.heuristics_->heuristics().size()) {
+    return false;
+  }
+
+  auto& heuristics = heuristics_->heuristics();
+  return std::equal(
+      heuristics.begin(),
+      heuristics.end(),
+      other.heuristics_->heuristics().begin(),
+      [](const SchedulerEntryPtr& a, const SchedulerEntryPtr& b) {
+        return a->sameAs(b.get());
+      });
+}
+
+void FusionSegmentRuntimeCache::evictId(size_t input_id) {
+  TORCH_INTERNAL_ASSERT(id_to_rt_.count(input_id) != 0);
+
+  // Evict the stored input tensor meta data
+  //  corresponding to input_id
+  id_to_rt_.at(input_id)->evictCache(input_id);
+  id_to_rt_.erase(input_id);
+}
+
+FusionSegmentRuntime* FusionSegmentRuntimeCache::getRTById(size_t input_id) {
+  if (id_to_rt_.count(input_id) == 0) {
+    return nullptr;
+  }
+  return id_to_rt_.at(input_id);
+}
+
+FusionSegmentRuntime* FusionSegmentRuntimeCache::getRTByHeuristics(
+    const at::ArrayRef<IValue>& inputs,
+    size_t input_id) {
+  auto dev_id = getCommonDeviceCUDA(inputs);
+  auto heuristics = segmented_fusion_->makeHeuristics(inputs);
+  EntryTag tag(heuristics.get());
+  auto rt = at(dev_id, tag);
+
+  // Heuristics miss
+  if (rt == nullptr) {
+    // Construct new runtime instance
+    auto new_rt = std::make_unique<FusionSegmentRuntime>(
+        segmented_fusion_, heuristics, input_id);
+    rt = new_rt.get();
+
+    // Cache the new instance
+    insertEntry(dev_id, tag, std::move(new_rt));
+  }
+
+  // Cache this new id
+  id_to_rt_[input_id] = rt;
+
+  return rt;
+}
+
+void FusionSegmentRuntimeCache::initCache(SegmentedFusion* sf) {
+  segmented_fusion_ = sf;
+}
+
+FusionSegmentRuntime* FusionSegmentRuntimeCache::at(int dev_id, EntryTag tag) {
+  if (seg_runtime_cache_group_.count(dev_id) == 0) {
```

Comment:
nitpick: avoid repetitive hashing.

Reply:
Great idea, thanks! Refactored `FusionSegmentRuntimeCache::at` and `FusionSegmentRuntimeCache::insertEntry`. I always worry about making code confusing to read when optimizing container type access. Please let me know if you have further suggestions on the refactored version.